File: ios/Packages/GemstoneServices/Sources/Stores/NodeStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.Chain
import protocol Gemstone.GemNodeStore
import struct Gemstone.Node
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneNodeStore: GemNodeStore, Sendable {
    private let store: NodeStore

    public init(store: NodeStore) {
        self.store = store
    }

    public func getNodes(chain: Gemstone.Chain) async throws -> [Gemstone.Node] {
        try store.nodes(chain: Primitives.Chain(id: chain)).map { $0.node.toGem() }
    }

    public func addNode(chain: Gemstone.Chain, node: Gemstone.Node) async throws {
        try store.addNodes(chainNodes: [ChainNodes(chain: chain, nodes: [node.toPrimitives()])])
    }

    public func deleteNode(chain: Gemstone.Chain, url: String) async throws {
        try store.deleteNode(chain: Primitives.Chain(id: chain), url: url)
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/NodeStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.service.store.database.NodesDao
import com.gemwallet.android.data.service.store.database.entities.DbNode
import com.gemwallet.android.ext.requireChain
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.Node
import uniffi.gemstone.GemNodeStore

class GemstoneNodeStore(private val nodesDao: NodesDao) : GemNodeStore {

    override suspend fun getNodes(chain: String): List<uniffi.gemstone.Node> = nodesDao.getNodes(chain.requireChain())
        .map { Node(it.url, it.status, it.priority).toGem() }

    override suspend fun addNode(chain: String, node: uniffi.gemstone.Node) {
        val value = node.toPrimitives()
        nodesDao.addNodes(listOf(DbNode(value.url, value.status, value.priority, chain.requireChain())))
    }

    override suspend fun deleteNode(chain: String, url: String) = nodesDao.deleteNode(chain.requireChain(), url)
}
